utils/difference.js

```diff
@@ -0,0 +1,61 @@
+const { executeWithValues: executeFormula } = require('./formula');
+
+/*
+ * Do calculations for all difference_* values for the given row
+ * @param{Object} row - The row to do calculations for
+ */
+function doDifferenceCalculations(row) {
+  calculateDifferences(row);
+  calculateDifferencePercents(row);
+}
+
+/*
+ * Calculate difference_sum, difference_m, and significant
+ * @param{row} - The row to do calculations for
+ */
+function calculateDifferences(row) {
+  const updatedRow = row;
+  if (exists(updatedRow.sum) && exists(updatedRow.comparison_sum)) {
+    updatedRow.difference_sum = executeFormula('delta', [updatedRow.sum, updatedRow.comparison_sum]);
+
+    if (exists(updatedRow.m) && exists(updatedRow.comparison_m)) {
+      updatedRow.difference_m = executeFormula('delta_m', [updatedRow.m, updatedRow.comparison_m]);
+
+      // TODO rename difference_significant
+      if (updatedRow.difference_sum !== 0) updatedRow.significant = executeFormula('significant', [updatedRow.difference_sum, updatedRow.difference_m]);
+    }
+  }
+
+  // special handling for 'decennial' rows, which do not have MOE and are all considered 'significant'
+  if (row.profile === 'decennial') row.significant = true;
+}
+
+/*
+ * Calculate difference_percent, difference_percent_m, and percent_significant
+ * @param{row} - The row to do calculations for
+ */
+function calculateDifferencePercents(row) {
+  const updatedRow = row;
+  if (exists(updatedRow.percent) && exists(updatedRow.comparison_percent)) {
+    const differencePercent = executeFormula('delta', [updatedRow.percent * 100, updatedRow.comparison_percent * 100]);
+    updatedRow.difference_percent = differencePercent < 0 && differencePercent > -0.05 ? 0 : differencePercent;
+
+    if (exists(updatedRow.percent_m) && exists(updatedRow.comparison_percent_m)) {
+      updatedRow.difference_percent_m = executeFormula('delta_m', [updatedRow.percent_m * 100, updatedRow.comparison_percent_m * 100]);
+
+      // TODO rename difference_percent_significant
+      if (updatedRow.difference_percent !== 0) updatedRow.percent_significant = executeFormula('significant', [updatedRow.difference_percent, updatedRow.difference_percent_m]);
+    }
+  }
+}
```

Comment:
Would these logical checks be appropriate as formulas too? Or does that shift code into more unfamiliar territory?

Reply:
these conditions can be added to the formulas if we think that's the correct implementation! 

I put them here because it saves us some computation -- we exit early & don't do any formula parsing at all if the values are not set. However, if it is clearer to have the boolean checks in the execl formulas, I'd say that's more important. I doubt this 'optimization' has any affect on the bottom line timing anyway.